receiver/receivercreator/factory.go

```diff
@@ -66,11 +70,79 @@ func createDefaultConfig() component.Config {
 	}
 }
 
+func createLogsReceiver(
+	_ context.Context,
+	params receiver.CreateSettings,
+	cfg component.Config,
+	consumer consumer.Logs,
+) (receiver.Logs, error) {
+	var err error
+	var recv receiver.Logs
+	rCfg := cfg.(*Config)
+	r := receivers.GetOrAdd(cfg, func() component.Component {
+		recv, err = newLogsReceiverCreator(params, rCfg, consumer)
+		return recv
+	})
+	rcvr := r.Component.(*receiverCreator)
+	if rcvr.nextLogsConsumer == nil {
+		rcvr.nextLogsConsumer = consumer
+	}
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
 func createMetricsReceiver(
-	ctx context.Context,
+	_ context.Context,
 	params receiver.CreateSettings,
 	cfg component.Config,
 	consumer consumer.Metrics,
 ) (receiver.Metrics, error) {
-	return newReceiverCreator(params, cfg.(*Config), consumer)
+	var err error
+	var recv receiver.Logs
+	rCfg := cfg.(*Config)
+	r := receivers.GetOrAdd(cfg, func() component.Component {
+		recv, err = newMetricsReceiverCreator(params, rCfg, consumer)
+		return recv
+	})
+	rcvr := r.Component.(*receiverCreator)
+	if rcvr.nextMetricsConsumer == nil {
+		rcvr.nextMetricsConsumer = consumer
+	}
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
 }
+
+func createTracesReceiver(
+	_ context.Context,
+	params receiver.CreateSettings,
+	cfg component.Config,
+	consumer consumer.Traces,
+) (receiver.Traces, error) {
+	var err error
+	var recv receiver.Logs
+	rCfg := cfg.(*Config)
+	r := receivers.GetOrAdd(cfg, func() component.Component {
+		recv, err = newTracesReceiverCreator(params, rCfg, consumer)
+		return recv
+	})
+	rcvr := r.Component.(*receiverCreator)
+	if rcvr.nextTracesConsumer == nil {
+		rcvr.nextTracesConsumer = consumer
+	}
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+var receivers = sharedcomponent.NewSharedComponents()
```

Comment:
nit: I wonder why these are defined at the bottom of the file (they are elsewhere too) 🤔

Reply:
moved them to the top (where they belong ;) )

Reply:
we're just been copying and pasting that code between components without much thought, good point